api/container.gql

```diff
@@ -22,17 +22,28 @@ Examples:
 scalar ContainerAddress
 
 """
-"An OCI-compatible container, also known as a "docker container"
+An OCI-compatible container, also known as a "docker container"
 """
 type Container {
     "A unique identifier for this container"
     id: ContainerID!
 
-    "Initialize this container from the base image published at the given address"
+    "Initialize this container to the base image published at the given address"
     from(address: ContainerAddress!): Container!
```

Comment:
[Bike shedding]

I see why it's this way, and the alternatives I have are not much better.

What bothers me is this chainable `from` is being more or less used as a "top level query with an empty selector" (e.g. `container(from: XXX)` vs `container { from(XXX) }`)

In practice, this is the only correct incantation:

```graphql
{
  container {
    from {
      ...
    }
}
```

Doing anything else, such as `{ container { withUser { from { ... } } }` doesn't work. This is because, even though `from` looks like it can be chained, it actually doesn't care whatsoever about the parent value (as per the buildkit API).

Perhaps we could do something custom to merge the internal representation of a container with the one coming from `llb.Image`, but that's a big if (what does it even mean to `container { withRoot(...) { from(...) { ... } }`?)

Alternatives:
- `container(id: ContainerID, from: String)` (both optional) and the user can decide to use one or another
  - However, that's weird. Can't find anything in the GitHub API that takes either THIS or THAT as selector for top-level queries
  - Even weirder in codegen
- Top level query
  - Lazy: `containerByID(id: ContainerID!): Container!` / `containerByImage(ref: String!): Container!` <-- ugly
  - Different selector: `container(id: ContainerID!): Container!` / `image(ref: String!): Container!` <-- why not ...

Reply:
This is a good summary of the pros and cons of this approach, and possible alternatives. I think it's the least bad so far.

> container(id: ContainerID, from: String) (both optional) and the user can decide to use one or another
> However, that's weird. Can't find anything in the GitHub API that takes either THIS or THAT as selector for top-level queries
> Even weirder in codegen

Yes, I agree it's weird. Feels like an implicit API and doesn't play to the strengths of GraphQL.

> Lazy: containerByID(id: ContainerID!): Container! / containerByImage(ref: String!): Container! <-- ugly

Yup, ugly :)

> Different selector: container(id: ContainerID!): Container! / image(ref: String!): Container! <-- why not ...

I don't like top-level `image` because it's not clear what kind of image we're talking about. IMO there should be a high bar of clarity and uniqueness to adding a top-level query field. "container" passes that bar but "image" doesn't, in my opinion.

Reply:
By the way here's another variation I tried:

```
// id may be one of:
//  1. ID computed from `id` field in a previous query
//  2. OCI-compatible remote image address
container(id: ContainerID)
```

Reply:
I think this is OK now that `withFS` has been added.

If it weren't there, I would agree it's a bit weird to have `container {}` and the only practical sub-query for it is `from`. But now that `withFS` is there, they both just become diifferent ways to set the rootfs. Only one sets just the rootfs, and the other sets the rootfs + config.

In fact even without `withFS` I can think of one way you could end up nesting `from`: you could start with a `/src` mount point containing your Go + JS app, then run `go build` to build your Go bits, and then `yarn build` to build your JS bits, all in one query:

```graphql

container {
  from(address: "golang:1.19") {
    withMountedDirectory(path: "/src", src: $repo) {
      withWorkdir(path: "/src") {
        exec(args: ["go", "build", "./cmd/cloak"]) {
          from(address: "node") {
            exec(args: ["yarn", "build"]) {
              directory(path: "/src") {
                id
              }
            }
          }
        }
      }
    }
  }
}
```

Is that a pattern we want to encourage? I dunno. But it'd be cool if it "just worked", and I _think_ it might.